dcpy/models/product/dataset/metadata_v2.py

```diff
@@ -0,0 +1,424 @@
+from __future__ import annotations
+
+import jinja2
+
+from pathlib import Path
+from pydantic import BaseModel, field_validator, model_serializer
+from pydantic.fields import PrivateAttr
+from typing import Any, List, Literal, get_args, Self
+import typing
+import unicodedata
+import yaml
+
+from dcpy.utils.logging import logger
+from dcpy.utils.collections import deep_merge_dict as merge
+
+
+# MISC UTILS
+def normalize_text(s):
+    """
+    Normalize the text we may receive from the various metadata sources.
+    Primarily useful for cleaning long-text like descriptions.
+    """
+    char_map = {
+        "–": "-",  # en dash to hyphen
+        "—": "-",  # em dash to hyphen
+        "’": "'",  # curly apostrophe
+        "“": '"',  # lcurly quote
+        "”": '"',  # rcurly quote
+    }
+    translator = str.maketrans(char_map)
+    # Normalize Unicode characters
+    normalized = unicodedata.normalize("NFKD", s)
+    # Apply the translation
+    cleaned = normalized.translate(translator)
+    return cleaned.strip()
+
+
+# BASE
+class SortedSerializedBase(BaseModel):
+    """A Pydantic BaseModel that will allow for sensible (and overrideable) deserialization order.
+
+    The serialization order is as follows:
+    - model attributes defined in the head sort order
+    - simple (and nullable simple) types: strings, ints, bools, literals
+    - complex types
+    - model attributes defined in the tail sort order
+
+    Note: This is put in its own class because it might be useful for other
+    classes unrelated to product metadata in the future.
+    """
+
+    _exclude_falsey_values: bool = True
+    _head_sort_order: list[str] = PrivateAttr(default=["id"])
+    _tail_sort_order: list[str] = PrivateAttr(default=["custom"])
+
+    @model_serializer(mode="wrap")
```

Comment:
just checking my/your understanding of a "wrap" serializer - this gets called on the pydantic object, returning some sort of dict/pydantic like object that we've configured as per this function, and _then_ its children are serialized? Or vice-versa -> handler gets us the "native" serialized dict, and we modify it there.

I guess those work out to roughly the same. Just find the pydantic docs a tad uncertain

Reply:
Yeah, the docs aren't great here. But to your question, the serialization is recursive - the model will descend through it's children attributes calling `model_dump` on those. That bubbles up to, eventually, just returning a top-level dictionary. Since all our metadata models inherit from this class, they'll all use this machinery to deserialize.